how to gracefully terminate a python node

I'm a little bit puzzled about the fact that the code in the finally branch has zero coverage. maybe rlcpy.ok() is always false here. Should the destroy_node() be called even if the context is not valid any more? @saikishor any idea?

![image](https://github.com/user-attachments/assets/fcf47e6b-6151-4c02-b0d2-afb894c00c9c)

_Originally posted by @christophfroehlich in https://github.com/ros-controls/ros2_controllers/issues/1369#issuecomment-2483034214_

This does not happen always, but if it enters there is a different error
```
2024-11-22T12:11:30.3702266Z [publisher_joint_trajectory_controller-2] Keyboard interrupt received. Shutting down node.
2024-11-22T12:11:30.3703121Z [publisher_joint_trajectory_controller-2] Traceback (most recent call last):
2024-11-22T12:11:30.3704880Z [publisher_joint_trajectory_controller-2]   File "/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/install/ros2_controllers_test_nodes/lib/ros2_controllers_test_nodes/publisher_joint_trajectory_controller", line 33, in <module>
2024-11-22T12:11:30.3706968Z [publisher_joint_trajectory_controller-2]     sys.exit(load_entry_point('ros2-controllers-test-nodes==4.16.0', 'console_scripts', 'publisher_joint_trajectory_controller')())
2024-11-22T12:11:30.3708246Z [publisher_joint_trajectory_controller-2]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-22T12:11:30.3710170Z [publisher_joint_trajectory_controller-2]   File "/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/install/ros2_controllers_test_nodes/lib/python3.12/site-packages/ros2_controllers_test_nodes/publisher_joint_trajectory_controller.py", line 196, in main
2024-11-22T12:11:30.3711758Z [publisher_joint_trajectory_controller-2]     rclpy.shutdown()
2024-11-22T12:11:30.3712778Z [publisher_joint_trajectory_controller-2]   File "/opt/ros/rolling/lib/python3.12/site-packages/rclpy/__init__.py", line 170, in shutdown
2024-11-22T12:11:30.3713790Z [publisher_joint_trajectory_controller-2]     _shutdown(context=context)
2024-11-22T12:11:30.9113697Z [publisher_joint_trajectory_controller-2]   File "/opt/ros/rolling/lib/python3.12/site-packages/rclpy/utilities.py", line 82, in shutdown
2024-11-22T12:11:30.9115644Z [publisher_joint_trajectory_controller-2]     context.shutdown()
2024-11-22T12:11:30.9117325Z [publisher_joint_trajectory_controller-2]   File "/opt/ros/rolling/lib/python3.12/site-packages/rclpy/context.py", line 179, in shutdown
2024-11-22T12:11:30.9118481Z [publisher_joint_trajectory_controller-2]     self._cleanup()
2024-11-22T12:11:30.9119840Z [publisher_joint_trajectory_controller-2]   File "/opt/ros/rolling/lib/python3.12/site-packages/rclpy/context.py", line 170, in _cleanup
2024-11-22T12:11:30.9121292Z [publisher_joint_trajectory_controller-2]     self.__context.shutdown()
2024-11-22T12:11:30.9122665Z [publisher_joint_trajectory_controller-2] rclpy._rclpy_pybind11.RCLError: failed to shutdown: rcl_shutdown already called on the given context, at ./src/rcl/init.c:290
2024-11-22T12:11:30.9125963Z [ERROR] [publisher_joint_trajectory_controller-2]: process has died [pid 30093, exit code 1, cmd '/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/install/ros2_controllers_test_nodes/lib/ros2_controllers_test_nodes/publisher_joint_trajectory_controller --ros-args --params-file /home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/install/ros2_controllers_test_nodes/share/ros2_controllers_test_nodes/test/
```
https://github.com/ros-controls/ros2_controllers/actions/runs/11972094365/job/33378268325?pr=1314

--- a/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_forward_position_controller.py
+++ b/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_forward_position_controller.py
@@ -77,7 +77,6 @@ def main(args=None):
     finally:
         if rclpy.ok():
             publisher_forward_position.destroy_node()
-            rclpy.shutdown()
 
 
 if __name__ == "__main__":

--- a/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_forward_position_controller.py
+++ b/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_forward_position_controller.py
@@ -70,13 +70,10 @@ def main(args=None):
 
     try:
         rclpy.spin(publisher_forward_position)
-    except KeyboardInterrupt:
+    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
         print("Keyboard interrupt received. Shutting down node.")
     except Exception as e:
         print(f"Unhandled exception: {e}")
-    finally:
-        if rclpy.ok():
-            publisher_forward_position.destroy_node()
 
 
 if __name__ == "__main__":

--- a/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_joint_trajectory_controller.py
+++ b/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_joint_trajectory_controller.py
@@ -186,13 +186,10 @@ def main(args=None):
 
     try:
         rclpy.spin(publisher_joint_trajectory)
-    except KeyboardInterrupt:
+    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
         print("Keyboard interrupt received. Shutting down node.")
     except Exception as e:
         print(f"Unhandled exception: {e}")
-    finally:
-        if rclpy.ok():
-            publisher_joint_trajectory.destroy_node()
 
 
 if __name__ == "__main__":

--- a/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_joint_trajectory_controller.py
+++ b/ros2_controllers_test_nodes/ros2_controllers_test_nodes/publisher_joint_trajectory_controller.py
@@ -193,7 +193,6 @@ def main(args=None):
     finally:
         if rclpy.ok():
             publisher_joint_trajectory.destroy_node()
-            rclpy.shutdown()
 
 
 if __name__ == "__main__":

